Improve resize transactions

kennylevinsen opened the pull request:
This PR contains a few improvements to the resizing logic:

1. Save all view buffers, including subsurfaces, rather than only the main surface. Fixes subsurfaces disappearing during resize.
2. Ensure that "current" geometry is not trashed, and is used when appropriate. Fixes border/titlebar and content disagreeing on dimensions during resize.
3. Update centering logic to account for complete and current dimensions. Fixes positioning during resize, especially with CSDs.
4. Remove bogus dimensions being tracked on the container state that was no longer useful.

kennylevinsen commented:
cc @emersion, @kchibisov

Note: A damage issue can currently be seen if one resizes e.g. firefox aggressively enough. Unsure if this is new, as firefox currently had nothing to render during resize, but worth a look regardless.

kchibisov commented:
Seems to fix https://github.com/swaywm/sway/issues/4441 and firefox flickering on resize.

kennylevinsen commented:
Another note: I am questioning the use of these saved buffers. This PR fixes them, but I fail to see a scenario where they *actually* help over just rendering the latest buffers, and not dealing with saved buffers also fixes the issue.

emersion commented:
>I am questioning the use of these saved buffers

This is used for atomic transactions. When re-configuring the layout we need to send new configure events to all clients. We don't update anything till all clients have ack'ed the configure event (with a timeout). When a client submits an updated buffer but another client is still pending, we need to render the previous buffer. That's why we save it.

kennylevinsen commented:
So the point is to not change anything till everything is ready? That would make sense, but not sure how effective this is at the moment. E.g., in local tests with a 2x2 grid, resizing with the mouse so that all views change size leads to buffers being all over the place, with no visible difference with and without saved buffers.

Have not debugged further.

emersion commented:
> Expected size in transactions are not accounting for borders/decorations, so they often assume the new buffer is bad.

That should only affect xwayland right? Wayland has a serial mechanism.

kennylevinsen commented:
I have managed to obtain pretty good resize behavior. I took the liberty of expanding the scope of the PR slightly to cover these.

kennylevinsen commented:
I think this is good enough for now. More improvements (e.g. moving pending states to state structs like is done with current state, maybe clarifying the different sizes) can be done in a later PR.

emersion commented on include/sway/tree/view.h:
```diff
@@ -80,8 +83,7 @@ struct sway_view {
 
-	struct wlr_client_buffer *saved_buffer;
-	int saved_buffer_width, saved_buffer_height;
+	list_t *saved_buffers;
```
Can we use `wl_list` instead of `list_t` here? It seems more appropriate.

kennylevinsen commented:
1. Converted `list_t` to `struct wl_list`
2. Fixed a multi-output bug where saved buffer rendering would stop on first failed output intersect instead of proceeding to the next buffer that might intersect
3. Added some cleanup on view_destroy that we were technically missing.
